xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,292 @@
+// symbol, count        XXXXXX║ A,2 ║XXXXX║ B,1 ║ B,1 ║ C,3 ║XXXXXXXXXXXXXXXX
+// (input channel)      ──────╨─────╨─────╨─────╨─────╨─────╨────────────────
+// last                                   ┌─────┐     ┌─────┐
+// (input channel)      ──────────────────┘     └─────┘     └────────────────
+//                      ╥─────╥───────────╥───────────╥─────────────────╥────
+// state.symbol         ║ 0   ║ A         ║ B         ║ C               ║ 0
+// (set state value)    ╨─────╨───────────╨───────────╨─────────────────╨────
+//                      ╥─────╥─────╥─────╥───────────╥─────╥─────╥─────╥────
+// state.count          ║ 0   ║ 1   ║ 0   ║ 0         ║ 2   ║ 1   ║ 0   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_common as rle_common
+
+type DecInData  = rle_common::EncData;
+type DecOutData = rle_common::PlainData;
+
+// structure to preserve the state of an RLE decoder
+struct RLEDecState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // symbol to be repeated on output
+    symbol: bits[SYMBOL_WIDTH],
+    // count of symbols that has to be send
+    count: bits[COUNT_WIDTH],
+    // send last when repeat ends
+    last: bool,
+}
+ // RLE decoder implementation
+pub proc RLEDec<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+    init {(
+        RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            symbol: bits[SYMBOL_WIDTH]:0,
+            count:  bits[COUNT_WIDTH]:0,
+            last:   bool:false,
+        }
+    )}
+
+    config (
+        input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+        output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+        let (tok, input) = recv_if(tok, input_r, recv_next_symbol, zero_input);
+        let (next_symbol, count, next_last) = if (recv_next_symbol) {
+            (input.symbol, input.count, input.last)
+        } else {
+            (state.symbol, state.count, state.last)
+        };
+        let next_count = count - bits[COUNT_WIDTH]:1;
+        let send_last = next_last && (next_count == bits[COUNT_WIDTH]:0);
+        let data_tok = send(tok, output_s, DecOutData {symbol: next_symbol, last: send_last});
+        if (send_last) {
```

Comment:
I wonder if we could be more explicit here by flattening the three possible state (with a match statement):
https://google.github.io/xls/dslx_reference/#match-expression
```
  match (recv_next_symbol, done_sending, send_last) {
    (_, true, true) => zero!<RLEDecState>(),
    (true, _, _) => (input.symbol, input.count, input.last)
    (false, _, _) => (state.symbol, state.count, state.last)
  }
```
wdy?

Reply:
I agree it will make this code cleaner.
I'll change it in next commit